# Exploring Ebay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. We've made a few modifications from the original dataset that was uploaded to Kaggle:

- We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
- We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

NOTES:
- Entries of object columns are often written in German
- **price** column has '$' symbol in front
- Dataframe has integer indexing

The columns in the dataframe are:

In [5]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Let's change them from camelcase to snakecase:

In [6]:
old_colnames = autos.columns.to_series()

col_dict = {
    "yearOfRegistration":"registration_year",
    "monthOfRegistration":"registration_month",
    "notRepairedDamage":"unrepaired_damage",
    "dateCreated":"ad_created",
    "dateCrawled":"date_crawled",
    "offerType":"offer_type",
    "vehicleType":"vehicle_type",
    "powerPS":"power_ps",
    "fuelType":"fuel_type",
    "nrOfPictures":"nr_of_pictures",
    "postalCode":"postal_code",
    "lastSeen":"last_seen"    
}

# Add old columns too
for c in old_colnames:
    if c not in col_dict:
        col_dict[c] = c
        

new_colnames = old_colnames.map( col_dict )
#print(new_colnames)

autos.columns = new_colnames

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We renamed columns from camelCase to snakecase (e.g. dateCreated --> date_created)

Let's take a look at the columns in our data:

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Notes:
- Columns **seller** and **offer_type** have almost always the same value (*privat* and *Angebot*, respectively). Should probably be dropped
- price and odometer are stored as strings but are essentially numeric

In [8]:
print(autos["price"].unique())
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].str.replace(".", "")
autos["price"] = autos["price"].astype(float)
print(autos["price"].describe())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [9]:
print(autos["odometer"].unique())
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")
autos["odometer"] = autos["odometer"].astype(float)

autos.rename( {"odometer":"odometer_km"}, axis=1, inplace=True )
print(autos["odometer_km"].describe())

['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


The average car is driven 15000 miles per year so we see these cars have been used quite a lot (~5 years of average use).

Let's explore odometer_km and price column data for fishy entries:

In [10]:
print(autos["price"].unique().shape)
print()
print(autos["price"].describe(include='all'))
print()
print(autos["price"].value_counts().sort_index().head())
print()
print(autos["price"].value_counts().sort_index(ascending=False).head())
print()
print(autos["price"].max())

(2357,)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

0.0    1421
1.0     156
2.0       3
3.0       1
5.0       2
Name: price, dtype: int64

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
Name: price, dtype: int64

99999999.0


<br>
NOTE:
- I'll remove all vehicles with price greater than 10000 (4 outliers) 
- I'll keep vehicles with price equal to zero as there are 1421 of them

In [11]:
autos = autos.loc[ autos["price"] < 10000 ]

In [12]:
print(autos["odometer_km"].unique().shape)
print()
print(autos["odometer_km"].describe())
print()
print(autos["odometer_km"].value_counts().sort_index().head())
print()
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head())
print()
print(autos["odometer_km"].max())

(13,)

count     41819.000000
mean     132205.098161
std       34807.939617
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

5000.0     804
10000.0     95
20000.0    321
30000.0    326
40000.0    347
Name: odometer_km, dtype: int64

150000.0    29992
125000.0     4249
100000.0     1602
90000.0      1254
80000.0       934
Name: odometer_km, dtype: int64

150000.0


<br>
Now let's have a look at the date columns **date_crawled**, **ad_created** and **last_seen**:

In [13]:
autos[ ["date_crawled", "ad_created", "last_seen"] ].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [22]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index().head()
)

2016-03-05    0.025276
2016-03-06    0.013774
2016-03-07    0.036228
2016-03-08    0.034243
2016-03-09    0.033406
Name: date_crawled, dtype: float64

In [23]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index().tail()
)

2016-04-03    0.038093
2016-04-04    0.036347
2016-04-05    0.013224
2016-04-06    0.003396
2016-04-07    0.001339
Name: date_crawled, dtype: float64

The date_crawled values range from March 5th to April 7th of 2016

In [26]:
(autos["ad_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index().head(10)
)

2015-08-10    0.000024
2015-09-09    0.000024
2015-11-10    0.000024
2015-12-05    0.000024
2015-12-30    0.000024
2016-01-07    0.000024
2016-01-10    0.000024
2016-01-13    0.000024
2016-01-16    0.000024
2016-01-27    0.000072
Name: ad_created, dtype: float64

In [27]:
(autos["ad_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index().tail(10)
)

2016-03-29    0.034219
2016-03-30    0.033597
2016-03-31    0.031899
2016-04-01    0.033310
2016-04-02    0.034099
2016-04-03    0.038332
2016-04-04    0.036897
2016-04-05    0.011861
2016-04-06    0.003419
2016-04-07    0.001243
Name: ad_created, dtype: float64

We see again the same time range. Interestingly the ad_created date distribution is heavier towards later dates.

In [29]:
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index().head()
)

2016-03-05    0.001148
2016-03-06    0.005046
2016-03-07    0.006074
2016-03-08    0.008441
2016-03-09    0.010713
Name: last_seen, dtype: float64

In [30]:
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index().tail()
)

2016-04-03    0.025945
2016-04-04    0.025682
2016-04-05    0.116885
2016-04-06    0.206413
2016-04-07    0.120854
Name: last_seen, dtype: float64

In [31]:
autos["registration_year"].describe()

count    41819.000000
mean      2003.984816
std        101.122142
min       1000.000000
25%       1999.000000
50%       2002.000000
75%       2006.000000
max       9999.000000
Name: registration_year, dtype: float64

We see there are entries from the year 1000 and 9999 so that's clearly wrong. We should either drop those rows or replace them with a more reasonable value.

In [37]:
autos.loc[autos["registration_year"].between(1950, 2018), "registration_year" ].shape

(41779,)

The above shows 41779 out of 41819 non-null entries (99.9%) are between 1950 and 2018.

Obviously any values beyond 2016 (the latest date in **last_seen**), or at least beyond the current year (2018), have to be wrong:

In [67]:
autos.loc[autos["registration_year"]>2016, "registration_year" ].shape

(1820,)

In [65]:
autos.loc[autos["registration_year"]>2017, "registration_year" ].shape

(472,)

In [66]:
autos.loc[autos["registration_year"]>2018, "registration_year" ].shape

(18,)

In [45]:
autos.loc[autos["registration_year"]<1900, "registration_year" ]

10556    1800
22316    1000
24511    1111
32585    1800
35238    1500
49283    1001
Name: registration_year, dtype: int64

Let's count how many cars are in 2016 (supposedly the last date of this dataset), 2017, 2018 and beyond (which is impossible as that is the current year of writing of this notebook).

In [72]:
sum(autos["registration_year"].between(0,2016))

39999

In [74]:
sum(autos["registration_year"]==2016)

1235

In [75]:
sum(autos["registration_year"]==2017)

1348

In [76]:
sum(autos["registration_year"]==2018)

454

In [78]:
sum(autos["registration_year"].between(2019,99999))

18

We see a sharp fall beyond 2018 which leads us to believe maybe values up to 2018 are indeed ok. Let's drop anything before 1900 and after 2018:

In [79]:
autos = autos[autos["registration_year"].between(1900,2018)]

In [83]:
(autos["registration_year"]
 .value_counts(normalize=True)
 .sort_index(ascending=False).head(10))

2018    0.010863
2017    0.032253
2016    0.029549
2015    0.001531
2014    0.002560
2013    0.003972
2012    0.009905
2011    0.013590
2010    0.018064
2009    0.031176
Name: registration_year, dtype: float64

Now let's take a look at the **brand** column:

In [92]:
all_brands = autos["brand"].unique()
print("Unique brands in autos dataset:\n{}".format(all_brands))

Unique brands in autos dataset:
['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'opel' 'mazda' 'mini' 'toyota' 'dacia' 'nissan'
 'jeep' 'saab' 'volvo' 'mitsubishi' 'fiat' 'skoda' 'subaru'
 'sonstige_autos' 'kia' 'citroen' 'porsche' 'hyundai' 'chevrolet' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'jaguar' 'alfa_romeo' 'lada'
 'rover' 'daihatsu' 'lancia']


Top 20 brands:

In [157]:
print(autos["brand"].value_counts().head(10))

volkswagen       9037
opel             5202
bmw              3948
mercedes_benz    3436
ford             3188
audi             2907
renault          2332
peugeot          1401
fiat             1268
seat              855
Name: brand, dtype: int64


In [158]:
top_brands = autos["brand"].value_counts().sort_values().index
print(autos["brand"].value_counts().sort_values(ascending=False).head(10))

volkswagen       9037
opel             5202
bmw              3948
mercedes_benz    3436
ford             3188
audi             2907
renault          2332
peugeot          1401
fiat             1268
seat              855
Name: brand, dtype: int64


In [159]:
mean_price_by_brand = {}
for b in top_brands:
    mean_price_by_brand[b] = (autos
                              .loc[ autos["brand"]==b, "price"]
                              .mean().round(2)
                             )

In [160]:
for key, value in sorted(mean_price_by_brand.items(), 
                         key=lambda kv: (kv[1],kv[0]), reverse=True):
    print("{}: {}".format(key,value))

mini: 6169.47
jeep: 4964.36
dacia: 4800.27
porsche: 4302.05
land_rover: 4280.62
skoda: 4279.26
bmw: 3964.25
toyota: 3962.77
mercedes_benz: 3788.53
chevrolet: 3771.9
audi: 3765.65
hyundai: 3643.15
kia: 3583.93
jaguar: 3440.5
smart: 3390.8
suzuki: 3259.28
sonstige_autos: 3103.99
seat: 3063.95
volkswagen: 3039.24
citroen: 2980.2
honda: 2919.01
mazda: 2873.47
volvo: 2765.84
alfa_romeo: 2741.23
nissan: 2702.38
peugeot: 2576.65
lada: 2476.9
fiat: 2363.58
ford: 2348.44
chrysler: 2343.33
mitsubishi: 2272.95
subaru: 2269.89
opel: 2253.64
saab: 2230.78
renault: 1943.03
trabant: 1557.21
daihatsu: 1552.09
lancia: 1551.15
rover: 1170.91
daewoo: 1038.35


We see that *mini* is the brand with highest mean price in our dataset, while *daewoo* has the lowest. The most popular brand *volkswagen* sits comfortably in the middle price range. 

Let's take a closer look at the 6 top brands: *bmw*, *mercedes_benz*, *opel*, *audi*, *volkswagen*, *ford*.

In [167]:
top_brands = (autos["brand"].value_counts()
              .sort_values(ascending=False).head(6).index)
print(top_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'ford', 'audi'], dtype='object')


In [169]:
bmp_dict = {}
for b in top_brands:
    bmp_dict[b] = (autos
                   .loc[ autos["brand"]==b, "price"]
                   .mean().round(2))

In [178]:
bmm_dict = {}
for b in top_brands:
    bmm_dict[b] = (autos
                   .loc[ autos["brand"]==b, "odometer_km"]
                   .mean().round(2))
print(bmm_dict)

{'mercedes_benz': 140708.67, 'opel': 132572.09, 'ford': 128861.36, 'audi': 143942.21, 'bmw': 141161.35, 'volkswagen': 136431.34}


In [180]:
bmp_series = pd.Series(bmp_dict)
bmm_series = pd.Series(bmm_dict)

In [188]:
df = pd.DataFrame(bmp_series, columns=["mean_price"])
df["mean_mileage"] = bmm_series

df.head()

mean_price  mean_mileage
audi              3765.65     143942.21
bmw               3964.25     141161.35
ford              2348.44     128861.36
mercedes_benz     3788.53     140708.67
opel              2253.64     132572.09

In [189]:
df.describe()

mean_price   mean_mileage
count     6.000000       6.000000
mean   3193.291667  137279.503333
std     761.147948    5743.933081
min    2253.640000  128861.360000
25%    2521.140000  133536.902500
50%    3402.445000  138570.005000
75%    3782.810000  141048.180000
max    3964.250000  143942.210000

We see that all brands have mean_mileage above the average of all brands (~126000km)

## Next steps if interested:

- Data cleaning:

    Identify categorical data that uses german words, translate them and map the values to their english counterparts
Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
See if there are particular keywords in the name column that you can extract as new columns

- Analysis:

    Find the most common brand/model combinations
Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
How much cheaper are cars with damage than their non-damaged counterparts?